geopandas/base.py

```diff
@@ -4104,6 +4104,103 @@ def sample_points(self, size, method="uniform", seed=None, **kwargs):
 
         return GeoSeries(result, name="sampled_points", crs=self.crs, index=self.index)
 
+    def polygonize(self, node=True, full=False):
+        """Creates polygons formed from the linework of a GeoSeries.
+
+        Polygonizes the GeoSeries that contain linework which represents the
+        edges of a planar graph. Any geometry type may be provided as input; only the
+        constituent lines and rings will be used to create the output polygons.
+
+        Lines or rings that when combined do not completely close a polygon will result
+        in an empty GeoSeries. Duplicate segments are ignored.
+
+        Unless you know that the input GeoSeries represent clean topological of a planar
+        graph (e.g. there is a node on both lines where they intersect), it is
+        recommended to use ``node=True`` which performs noding prior polygonization.
+        Using ``node=False`` will provide performance benefits but may result in
+        incorrect polygons if the input is not of the proper topology.
+
+        When ``full=True``, return not only the polyoginal result all extra outputs as
+        well. The return value consists of 4 elements:
+
+            - GeoSeries of the valid polygons (same as with ``full=False``)
+            - GeoSeries of cut edges: edges connected on both ends but not part of
+              polygonal output
+            - GeoSeries of dangles: edges connected on one end but not part of polygonal
+              output
+            - GeoSeries of invalid rings: polygons that are formed but are not valid
+              (bowties, etc)
+
+        Parameters
+        ----------
+        node : bool, optional
+            perform noding prior polygonization, by default True
+        full : bool, optional
+            return the full output composed of a tuple of GeoSeries, by default False
+
+        Returns
+        -------
+        GeoSeries | tuple(GeoSeries, GeoSeries, GeoSeries, GeoSeries)
+            GeoSeries with the polygons or a tuple of four GeoSeries as
+            ``(polygons, cuts, dangles, invalid)``
+
+        Examples
+        --------
+        >>> from shapely.geometry import LineString
+        >>> s = geopandas.GeoSeries([
+        ...     LineString([(0, 0), (1, 1)]),
+        ...     LineString([(0, 0), (0, 1), (1, 1), (1, 0), (0, 0)]),
+        ... ])
+        >>> s.polygonize()
+        0    POLYGON ((0.00000 0.00000, 1.00000 1.00000, 1....
+        1    POLYGON ((1.00000 1.00000, 0.00000 0.00000, 0....
+        Name: polygons, dtype: geometry
```

Comment:
I don't know how easy it is to construct but could we have an example where the number of rows in the input is different to the number in the output? (perhaps with a duplicate or an invalid linestring)

Just to make it obvious that all the rows is treated as the input to polygonize, rather than polygonize being called separately on each row (as we support geometrycollection as a geometry type).

(perhaps the first line of the docstring, and the fact there is no geodataframe method makes this obvious enough?)